File: common/src/main/java/dev/mayaqq/estrogen/client/registry/trinkets/EstrogenPatchesRenderer.java
package dev.mayaqq.estrogen.client.registry.trinkets;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Axis;
import dev.mayaqq.estrogen.config.EstrogenConfig;
import dev.mayaqq.estrogen.registry.EstrogenItems;
import earth.terrarium.baubly.client.BaubleRenderer;
import earth.terrarium.baubly.client.BaublyClient;
import earth.terrarium.baubly.common.SlotInfo;
import net.minecraft.client.Minecraft;
import net.minecraft.client.model.EntityModel;
import net.minecraft.client.model.PlayerModel;
import net.minecraft.client.model.geom.ModelPart;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.ItemDisplayContext;
import net.minecraft.world.item.ItemStack;

public class EstrogenPatchesRenderer implements BaubleRenderer {

    public EstrogenPatchesRenderer() {}

    @Override
    public void render(
            ItemStack stack,
            SlotInfo slotContext,
            PoseStack matrixStack,
            EntityModel<? extends LivingEntity> entityModel,
            MultiBufferSource renderTypeBuffer,
            int light,
            float limbSwing,
            float limbSwingAmount,
            float partialTicks,
            float ageInTicks,
            float netHeadYaw,
            float headPitch
    ) {
        if (entityModel instanceof PlayerModel<? extends LivingEntity> playerModel) {
            if (!EstrogenConfig.client().estrogenPatchRender.get()) return;
            LocalPlayer player = Minecraft.getInstance().player;
            matrixStack.pushPose();
            ModelPart leg = playerModel.leftLeg;

            // Sync with the player's animation
            if (player.isCrouching() && !playerModel.riding && !player.isSwimming()) {
                matrixStack.translate(0.0F, 0.0F, 0.25F);
            }
            matrixStack.translate(0.125F, 0.75F, 0.0F);
            matrixStack.mulPose(Axis.ZP.rotation(leg.zRot));
            matrixStack.mulPose(Axis.YP.rotation(leg.yRot));
            matrixStack.mulPose(Axis.XP.rotation(leg.xRot));
            matrixStack.translate(0.0F, 0.75F, 0.0F);

            // Rotate the patch to be on the player's thigh side
            matrixStack.mulPose(Axis.YN.rotation(89.6F));

            // move the patch to be on the player's thigh
            matrixStack.translate(0.0F, -0.6F, -0.135F);

            // Make the patch smaller
            matrixStack.scale(0.3F, 0.3F, 0.3F);

            Minecraft.getInstance().getItemRenderer().renderStatic(stack, ItemDisplayContext.FIXED, light, light, matrixStack, renderTypeBuffer, player.level(), 0);
            matrixStack.popPose();
        }
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/client/content/baubles/EstrogenPatchesRenderer.kt
package dev.mayaqq.estrogen.client.content.baubles

import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.math.Axis
import dev.mayaqq.cynosure.client.utils.pushPop
import dev.mayaqq.estrogen.config.EstrogenClientConfig
import earth.terrarium.baubly.client.BaubleRenderer
import earth.terrarium.baubly.common.SlotInfo
import net.minecraft.client.Minecraft
import net.minecraft.client.model.EntityModel
import net.minecraft.client.model.PlayerModel
import net.minecraft.client.model.geom.ModelPart
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.item.ItemDisplayContext
import net.minecraft.world.item.ItemStack
//TODO: Implement this on the item
class EstrogenPatchesRenderer : BaubleRenderer {
    override fun render(
        stack: ItemStack,
        slotInfo: SlotInfo,
        pose: PoseStack,
        model: EntityModel<out LivingEntity>,
        buffer: MultiBufferSource,
        light: Int,
        limbSwing: Float,
        limbSwingAmount: Float,
        partialTicks: Float,
        ageInTicks: Float,
        netHeadYaw: Float,
        headPitch: Float
    ) {
        if (model is PlayerModel<out LivingEntity?>) {
            if (!EstrogenClientConfig.Accessories.renderEstrogenPatches) return
            val player = Minecraft.getInstance().player ?: return
            pose.pushPop {
                val leg: ModelPart = model.leftLeg

                // Sync with the player's animation
                if (player.isCrouching && !model.riding && !player.isSwimming) {
                    translate(0.0f, 0.0f, 0.25f)
                }
                translate(0.125f, 0.75f, 0.0f)
                mulPose(Axis.ZP.rotation(leg.zRot))
                mulPose(Axis.YP.rotation(leg.yRot))
                mulPose(Axis.XP.rotation(leg.xRot))
                translate(0.0f, 0.75f, 0.0f)

                // Rotate the patch to be on the player's thigh side
                mulPose(Axis.YN.rotation(89.6f))

                // move the patch to be on the player's thigh
                translate(0.0f, -0.6f, -0.135f)

                // Make the patch smaller
                scale(0.3f, 0.3f, 0.3f)

                Minecraft.getInstance().itemRenderer.renderStatic(
                    stack,
                    ItemDisplayContext.FIXED,
                    light,
                    light,
                    pose,
                    buffer,
                    player.level(),
                    0
                )
            }
        }
    }
}